app/com/arpnetworking/metrics/portal/alerts/impl/DatabaseAlertRepository.java

```diff
@@ -0,0 +1,365 @@
+                    .setOperator(ebeanAlert.getOperator())
+                    .setPeriod(Period.seconds(ebeanAlert.getPeriod()))
+                    .setService(ebeanAlert.getService())
+                    .setStatistic(ebeanAlert.getStatistic())
+                    .setValue(new DefaultQuantity.Builder()
+                            .setValue(ebeanAlert.getQuantityValue())
+                            .setUnit(ebeanAlert.getQuantityUnit())
+                            .build())
+                    .setExtensions(mergeExtensions(ebeanAlert.getNagiosExtension()))
+                    .build();
+            values.add(result);
+        });
+
+        // Transform the results
+        return new DefaultQueryResult<>(values, pagedAlerts.getTotalRowCount(), etag.toString());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long getAlertCount() {
+        assertIsOpen();
+        return Ebean.find(models.ebean.Alert.class)
+                .findRowCount();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addOrUpdateAlert(final Alert alert) {
+        assertIsOpen();
+        LOGGER.debug()
+                .setMessage("Upserting alert")
+                .addData("alert", alert)
+                .log();
+
+
+        try (Transaction transaction = Ebean.beginTransaction()) {
+            models.ebean.Alert ebeanAlert = Ebean.find(models.ebean.Alert.class)
+                    .where()
+                    .eq("uuid", alert.getId())
+                    .findUnique();
+            boolean created = false;
+            if (ebeanAlert == null) {
+                ebeanAlert = new models.ebean.Alert();
+                created = true;
+            }
+
+            ebeanAlert.setCluster(alert.getCluster());
+            ebeanAlert.setUuid(alert.getId());
+            ebeanAlert.setMetric(alert.getMetric());
+            ebeanAlert.setContext(alert.getContext());
+            ebeanAlert.setNagiosExtension(getNagiosExtensionFromMap(alert.getExtensions()));
+            ebeanAlert.setName(alert.getName());
+            ebeanAlert.setOperator(alert.getOperator());
+            ebeanAlert.setPeriod(alert.getPeriod().toStandardSeconds().getSeconds());
+            ebeanAlert.setQuantityValue(alert.getValue().getValue());
+            ebeanAlert.setQuantityUnit(alert.getValue().getUnit().isPresent() ? alert.getValue().getUnit().get() : null);
+            ebeanAlert.setStatistic(alert.getStatistic());
+            ebeanAlert.setService(alert.getService());
+            _alertQueryGenerator.saveAlert(ebeanAlert);
+            transaction.commit();
+
+            LOGGER.info()
+                    .setMessage("Upserted alert")
+                    .addData("alert", alert)
+                    .addData("isCreated", created)
+                    .log();
+        } catch (final IOException e) {
+            LOGGER.error()
+                    .setMessage("Failed to close the transaction")
```

Comment:
Are you sure that failing to close the transaction means that the commit was successful?

In general since the catch blocks are identical the only difference is in the stack traces which should reflect the difference you describe. Otherwise, it sounds like you are telling me that one exception should be propagated because its a failure and the other is not a failure - then why is it propagated and moreover why is it even an exception? -- my point is not to not propagate or not catch, but evidence that your assumption may not be correct.